changefeedccl: replace errors.Is with direct comparison for sentinel errors

isaactwong opened the pull request:
Fixes #154780

This PR improves changefeed performance by replacing expensive `errors.Is()` calls with a direct DecodeStatus enum in the event processing hot path.

## Background

Previously, the changefeed decoder used sentinel errors (`ErrUnwatchedFamily` and `ErrTableOffline`) to signal skip conditions. These were checked using `errors.Is()`, which walks the entire error chain performing type assertions and comparisons at each level.

## Changes

This PR replaces the sentinel error pattern with an explicit `DecodeStatus` enum that is returned as a separate value alongside the decoded row and error. The DecodeStatus has three states:
- `DecodeOK`: successful decode
- `DecodeSkipUnwatchedFamily`: KV is from an unwatched column family
- `DecodeSkipTableOffline`: KV is from an offline table

This approach provides several benefits:
1. **Performance**: No error chain traversal overhead
2. **Safety**: Immune to error wrapping issues
3. **Clarity**: Explicitly documents skip vs error conditions
4. **Type-safe**: Enum prevents invalid status values
5. **Maintainability**: Easy to extend with new skip conditions

Changes include:
- Updated Decoder interface to return `(Row, DecodeStatus, error)`
- Updated all implementations and call sites across changefeed and cross-cluster replication code
- Updated tests to check status instead of sentinel errors
- Added explicit comments documenting resource cleanup behavior

## Performance Impact

Benchmark results demonstrate measurable improvements in the hot path:

```
OLD CODE (using errors.Is):
  watched-family:   766.2 ns/op   874 B/op   10 allocs/op
  unwatched-family: 423.0 ns/op  1192 B/op    6 allocs/op
  mixed-workload:   595.2 ns/op  1085 B/op    7 allocs/op

NEW CODE (using DecodeStatus enum):
  watched-family:   763.3 ns/op   874 B/op   10 allocs/op
  unwatched-family: 403.8 ns/op  1192 B/op    6 allocs/op
  mixed-workload:   551.3 ns/op  1086 B/op    7 allocs/op

IMPROVEMENTS:
  unwatched-family: 4.5% faster (19ns saved per event)
  mixed-workload:   7.4% faster (44ns saved per event)
```

This is particularly beneficial for high-throughput changefeeds where users watch only specific column families. In such cases, events for unwatched families are frequent and hit this code path on every event.

## Testing

- Existing tests updated to check status instead of errors.Is()
- Added comprehensive benchmarks in `event_bench_test.go`
- All tests continue to pass

Release note (performance improvement): Improved changefeed performance when filtering unwatched column families and offline tables by replacing expensive error chain traversal with direct status enum comparisons.

asg0451 reviewed (changes requested):
Thanks for the contribution Isaac & Claude! I have a qualm about the approach:

The danger of using `if err == sentinel` over `errors.Is()` is that somewhere in the (4?) layers between this callsite and where the error comes from, we may at any time add an `errors.Wrap()`. Doing so is ok in almost all cases. 

I think we have 2 options: either propagate an additional bool return variable through these sites and check that explicity, or add doc comments and unit tests to cement that returning a bare `ErrUnwatchedFamily | etc` is part of the api of these functions. The current unit tests test the `errors.Is`-ness of the return values only.

asg0451 commented:
Also, dont forget to sign the CLA :)

asg0451 commented on pkg/ccl/changefeedccl/cdcevent/event.go:
```diff
@@ -639,31 +649,35 @@ func (d *eventDecoder) DecodeKV(
-		return Row{}, err
+) (Row, DecodeStatus, error) {
+	status, err := d.initForKey(ctx, kv.Key, schemaTS, keyOnly)
+	if err != nil {
```
Won't this catch the `Row{}, DecodeSkip.., nil` case too? in which case we don't need the following if stmt.

asg0451 reviewed (approved):
Looks good to me. Totally optional, but it might be cool to add a benchmark for this code path so we can see what kind of improvement we get.

asg0451 commented:
Probably a roachperf benchmark would be most appropriate. I'd probably add a new test in https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/tests/cdc_bench.go like "cdc/scan/skip_lots_of_data" or similar. I'm not sure off the top of my head which workload uses column families heavily -- worst case we can just invent a workload that only writes to one family, and watch the other one.

asg0451 reviewed (approved):
Looks good to me, thanks for the contribution. I think the CI failures are infra flakes, I'm gonna poke them and check back later.

isaactwong committed:
changefeedccl: replace sentinel errors
with DecodeStatus enum

  This PR improves changefeed performance by replacing expensive
  errors.Is() calls with a direct DecodeStatus enum in the event
  processing hot path.

  The DecodeStatus enum has three states:
  - DecodeOK: successful decode
  - DecodeSkipUnwatchedFamily: KV is from an unwatched column family
  - DecodeSkipTableOffline: KV is from an offline table

  Benefits:
  1. Performance: No error chain traversal overhead
  2. Safety: Immune to error wrapping issues
  3. Clarity: Explicitly documents skip vs error conditions

  Benchmark results show 4.5-7.4% improvement in the hot path.

  Release note (performance improvement): Improved changefeed performance
  when filtering unwatched column families and offline tables by replacing
  expensive error chain traversal with direct status enum comparisons.

  Epic: None
